# remove trailing Ns and sequences with too many ambuiguities or too short
library(Biostrings)
library(tidyverse)
# function remove trailing Ns
source("code/functions/noTrailingN.R")
source("code/functions/ambiguityfilter.R")

# mitos assembled by me bwa mapped to references generated by iterative mapping in Geneious
path2itmitos <-
  "data/intermediate/consensus/bwa2iterative.fasta"
itmitos <-
  Biostrings::readDNAStringSet(path2itmitos) %>%
  noTrailingN()
# change names
names(itmitos) <-
  names(itmitos) %>%
  str_remove("_.*$")
# create data frame with AGTC frequencies
itmitos_fq <-
  Biostrings::oligonucleotideFrequency(itmitos, 1) %>%
  as.data.frame() %>%
  rowwise() %>%
  mutate(nonAmb = sum(across(A:`T`)))
itmitos_fq$code <- names(itmitos)
add_row(itmitos_fq,
            A = 0, C = 0, G = 0, T = 0, nonAmb = 0, code = "ZRC45046", .before = T) %>%
write_csv(file = "output/iterative_mitos_Nuc_fq.csv")

# filter incomplete
itmitos_filt <-
  itmitos %>%# remove trailing Ns
  {# remove short sequences
    .[width(.) > 10000]
  } %>%
  ambiguityfilter(maxpropAmb = .7)

# read meta
meta <-
  readRDS("data/intermediate/seqmeta.rds") %>%
  filter(sample %in% names(itmitos_filt)) %>%
  select(sample, taxa_new_taxonomy) %>%
  distinct() %>%
  mutate(seq_name = paste(str_replace(taxa_new_taxonomy, " ", "_"), sample, sep = "_"))

# complete names mitos. Add species
names(itmitos_filt) <-
  names(itmitos_filt) %>%
   sapply(function(x){
     if(str_length(x) > 10)
       stop("sequence names seem already complete")
     filter(meta, sample == x) %>%
       pull(seq_name)
   })

# mitos that Arlo assembled
path2arlo <-
  "data/intermediate/consensus/mitos_aligned_arlo.fasta"
toRetain <- c("ylomys")
##
arlo_mitos <-
  Biostrings::readDNAStringSet(path2arlo) %>%
  noTrailingN()
# create data frame with AGTC frequencies
arlo_mitos_fq <-
  Biostrings::oligonucleotideFrequency(arlo_mitos, 1) %>%
  as.data.frame() %>%
  rowwise() %>%
  mutate(nonAmb = sum(across(A:`T`)))
arlo_mitos_fq$code <- names(arlo_mitos)
write_csv(arlo_mitos_fq,
          file = "output/arlo_mitos_Nuc_fq.csv")


arlo_mitos_filt <-
  arlo_mitos %>% # remove trailing Ns
  ambiguityfilter(.7) %>%
  {# remove outgroups
    .[grep(pattern = toRetain, names(.))]
  }

#merge consensus seqs
allmitos <- c(itmitos_filt, arlo_mitos_filt)
# write 2 file
Biostrings::writeXStringSet(allmitos,
                            filepath = "data/intermediate/consensus/clean_mitos.fasta")
 
